src/inverse.jl

```diff
@@ -62,6 +62,8 @@ An instance `NoInverse(f)` signifies that `inverse(f)` is not defined.
 """
 struct NoInverse{F}
     f::F
+    NoInverse{F}(f) where F = new{F}(f)
+    NoInverse(f) = new{Core.Typeof(f)}(f)
```

Comment:
Why is this an inner constructor? Wouldn't it be sufficient to add this as an outer constructor and remove the default inner constructor in the line above?

Reply:
I took the way `Base.ComposedFunction` handles it as "best practice" here.

Reply:
Yes, I saw the implementation there. But it's not clear to me why an outer constructor is not sufficient and I would like to understand that 😅

Reply:
We can't do

```julia
struct NoInverse{F}
    f::F
end
export NoInverse
NoInverse(f) = NoInverse{Core.Typeof(f)}(f)
```

as it results in

```
WARNING: Method definition (::Type{InverseFunctions.NoInverse{F} where F})(Any) in module InverseFunctions at /user/.julia/dev/InverseFunctions/src/inverse.jl:64 overwritten at /user/.julia/dev/InverseFunctions/src/inverse.jl:68.
```

Reply:
I'm always confused what the default conatructors are and when they are omitted 😅 But I guess it would work (the outer constructor) if we add the standard inner constructor (as in the PR)?

Reply:
Is there a general guideline for when regular constructors (without type parameters) should be defined as inner constructors? I think I tend to define them as outer constructors because to me they look like outer constructors.

Reply:
I have no idea :-) ... this does work

```julia
struct NoInverse{F}
    f::F
    NoInverse{F}(f) where F = new{F}(f)
end
export NoInverse
NoInverse(f) = NoInverse{Core.Typeof(f)}(f)
```

I'm not entirely sure if there's any difference (for some compiler internals or whatever) to having both ctors inside.

Reply:
But I guess going with what Base does can't be too bad?

Reply:
I'd like to understand it first - I still think this is a prototypical example of an outer constructor and would implement it as an outer constructor.

To me the only reason to define such a constructor as an inner constructor would be the example in https://docs.julialang.org/en/v1/manual/constructors/#Outer-only-constructors - but since the default inner constructor is defined here as well the example and its motivation don't apply here.

Reply:
I'm with you, it'd like to understand that better myself. I asked on the Julia Slack, but as you maybe saw that wasn't very successful.

Reply:
According to Valentin (NOT vchuravy) (don't know his GitHub name), there is no difference to between having the other ctor inside or outside.

Reply:
I think that's @seelengrab? Unless I'm confused, which is likely

Reply:
> I'm always confused what the default conatructors are and when they are omitted 😅

The default constructors are removed when any kind of non-field block is in the `struct` definition, so having any function definition (even one that doesn't use `new`) removes them (though IIRC the removal behavior is only actually documented for "functions", it being presumed that the author of the code in question provided themselves with all the constructors they need).

In general, the default constructors are:

> The default constructor is equivalent to writing your own inner constructor method that takes all of the object's fields as parameters (constrained to be of the correct type, if the corresponding field has a type), and passes them to new, returning the resulting object:

(from [here](https://docs.julialang.org/en/v1/manual/constructors/#man-inner-constructor-methods))

The default constructors for this struct in its definition looks like this:

```julia
struct NoInverse{F}
    f::F
    NoInverse(f::F) where F = new(f)
end
```

which in turn calls the intrinsic `%new(NoInverse{F}, f)`  to actually construct the object.

Reply:
> I think that's @Seelengrab? Unless I'm confused, which is likely

Thanks - sorry @Seelengrab , I wanted to tag you, but forgot your handle.

Reply:
Ok @devmotion , so what do we do with the ctors? Inside or outside? :-)

Reply:
I saw the discussion and the turn it took 😉 But I guess Valentin has a point about dispatching on `Type` being sufficient and less likely to break. So I guess my first preference would be to add outer constructors for `Type` (that would also mean we don't have to re-define the default inner constructor) and otherwise add an outer constructor with `Core.Typeof` (but then we have to re-define the inner constructor, so the first option seems safer/more stable and simpler - at least when dealing with a single argument).

Reply:
> Thanks - sorry @Seelengrab , I wanted to tag you, but forgot your handle.

No worries! I too would like to forget my handle :sob: 

> But I guess Valentin has a point about dispatching on `Type` being sufficient and less likely to break.

The one place where I think `Type` is extremely cumbersome is if you have an unknown number of arguments - but since Julia doesn't (currently/yet) have structs with a variable number of fields, if you reach that case, you're down in the land of recursion-on-tuple-types anyway :)

Reply:
Ok, I changed the dispatch (see latest commit).

Reply:
Good like this?